src/main/scala/io/iohk/ethereum/consensus/Consensus.scala

```diff
@@ -1,41 +1,187 @@
 package io.iohk.ethereum.consensus
 
-import io.iohk.ethereum.domain.Block
-import io.iohk.ethereum.ledger.BlockExecutionError.ValidationBeforeExecError
-import io.iohk.ethereum.ledger.BlockExecutionSuccess
+import akka.util.ByteString
+import io.iohk.ethereum.consensus.blocks.{BlockGenerator, TestBlockGenerator}
+import io.iohk.ethereum.consensus.validators.Validators
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.BlockExecutionError.{ValidationAfterExecError, ValidationBeforeExecError}
+import io.iohk.ethereum.ledger.Ledger.VMImpl
+import io.iohk.ethereum.ledger.{BlockExecutionError, BlockExecutionSuccess, BlockPreparator}
 import io.iohk.ethereum.nodebuilder.Node
-import io.iohk.ethereum.validators.BlockHeaderValidator
+import io.iohk.ethereum.utils.{BlockchainConfig, Logger}
+import org.spongycastle.util.encoders.Hex
 
 /**
- * Abstraction for a consensus protocol.
+ * Abstraction for a consensus protocol implementation.
  *
  * @see [[io.iohk.ethereum.consensus.Protocol Protocol]]
  */
-// FIXME Lot's of stuff to do...
 trait Consensus {
+  /**
+   * The type of configuration [[io.iohk.ethereum.consensus.FullConsensusConfig#specific specific]]
+   * to this consensus protocol implementation.
+   */
   type Config <: AnyRef /*Product*/
 
   def protocol: Protocol
 
   def config: FullConsensusConfig[Config]
 
-  /** Returns `true` if this is the standard Ethereum PoW consensus protocol (`ethash`). */
-  final def isEthash: Boolean = protocol.isEthash
+  /**
+   * This is the VM used to prepare and generate blocks.
+   */
+  def vm: VMImpl
+
+  /**
+   * Provides the set of validators specific to this consensus protocol.
+   */
+  def validators: Validators
+
+  /**
+   * This is used by the [[io.iohk.ethereum.consensus.Consensus#blockGenerator blockGenerator]].
+   */
+  def blockPreparator: BlockPreparator
+
+  /**
+   * Returns the [[io.iohk.ethereum.consensus.blocks.BlockGenerator BlockGenerator]]
+   * this consensus protocol uses.
+   */
+  def blockGenerator: BlockGenerator
 
   /**
    * Starts the consensus protocol on the current `node`.
    */
   def startProtocol(node: Node): Unit
 
+  /**
+   * Stops the consensus protocol on the current node.
+   * This is called internally when the node terminates.
+   */
   def stopProtocol(): Unit
 
+  // Note Ledger uses this in importBlock
+  def validateBlockBeforeExecution(
```

Comment:
maybe we can include this method in `Validators`?

Reply:
Yes, I think it is the right thing to do. I started moving this stuff outside consensus but there was too much to refactor for this round so I eventually [gave up](https://github.com/input-output-hk/mantis/pull/410/commits/d8bf2f32ec411dff0eaede837ced5f5c99ac6ecd#diff-8f88229ce5e1d25dc74496902accebacL13-L19).
How about moving both this method and also `validateBlockAfterExecution` in `Validators`?

Reply:
[Moved](https://github.com/input-output-hk/mantis/pull/410/commits/f07dabba8eca7eba850f1caf1035788b1b952111)